metadetect/lsst/metacal_exposures.py

```diff
@@ -0,0 +1,291 @@
+    """
+
+    if types is None:
+        types = DEFAULT_TYPES
+
+    cen, _ = get_integer_center(exp.getWcs(), exp.getBBox(), as_double=True)
+
+    # wcs = exp.getWcs()
+
+    gwcs = get_galsim_jacobian_wcs(exp=exp, cen=cen)
+    pixel = gwcs.toWorld(galsim.Pixel(scale=1))
+    pixel_inv = galsim.Deconvolve(pixel)
+
+    psf_image_array = get_psf_kernel_image(exp=exp, cen=cen)
+    psf_flux = psf_image_array.sum()
+
+    eimage = exp.image.array.copy()
+    if rot:
+        eimage = np.rot90(eimage, k=1)
+
+    # force double, *it matters*.  There are clear artifacts
+    # present when comparing to the regular metacal if float32 is
+    # used
+    image = galsim.ImageD(eimage, wcs=gwcs)
+    psf_image = galsim.ImageD(psf_image_array, wcs=gwcs)
+
+    # interpolated psf image and its deconvolution
+    psf_int = galsim.InterpolatedImage(psf_image, x_interpolant=INTERP)
+
+    # deconvolved galaxy image, psf+pixel removed
+    image_int_nopsf = galsim.Convolve(
+        galsim.InterpolatedImage(image, x_interpolant=INTERP),
+        galsim.Deconvolve(psf_int),
+    )
+
+    # TODO we may not need to deconvolve the pixel from the psf
+    # interpolated psf deconvolved from pixel.  This is what
+    # we dilate, shear, etc and reconvolve the image by
+    psf_int_nopix = galsim.Convolve([psf_int, pixel_inv])
+
+    gauss_psf = _get_gauss_target_psf(psf_int_nopix, flux=psf_flux)
+
+    dilation = 1.0 + 2.0*STEP
+    psf_dilated_nopix = gauss_psf.dilate(dilation)
+    psf_dilated = galsim.Convolve(psf_dilated_nopix, pixel)
+
+    psf_dilated_image = psf_image.copy()
+    psf_dilated.drawImage(image=psf_dilated_image, method='no_pixel')
+
+    mdict = {}
+    for shear_type in types:
+        mdict[shear_type] = _get_metacal_exp(
+            exp=exp, image=image, image_int_nopsf=image_int_nopsf,
+            psf_dilated=psf_dilated, psf_dilated_image=psf_dilated_image,
+            shear_type=shear_type, rot=rot,
+        )
+
+    return mdict
+
+
+def _get_metacal_exp(
+    exp, image, image_int_nopsf, psf_dilated, psf_dilated_image, shear_type,
+    rot,
+):
+    sexp = afw_image.ExposureD(exp, deep=True)
+
+    if shear_type == 'noshear':
+        obj = image_int_nopsf
+    else:
+        shear = get_shear_from_type(shear_type)
+        obj = image_int_nopsf.shear(shear)
+
+    shimage = image.copy()
+    galsim.Convolve(obj, psf_dilated).drawImage(
+        image=shimage, method='no_pixel',
+    )
+
+    simage = shimage.array
+    if rot:
+        simage = np.rot90(simage, k=3)
+
+    sexp.image.array[:, :] = simage
+
+    stack_psf = get_stack_kernel_psf(psf_dilated_image.array)
+    sexp.setPsf(stack_psf)
+    return sexp
+
+
+def get_shear_from_type(shear_type):
+    """
+    Convert a shear type string to a galsim Shear
+
+    Parameters
+    ----------
+    shear_type: str
+        One of ('noshear', '1p', '1m', '2p', '1m')
+
+    Returns
+    -------
+    galsim.Shear
+    """
+    if shear_type == '1p':
```

Comment:
check out the `_get_metacal_exp` function, it is dealt with there.  We don't actually do a shear for the noshear case because it would be wasted cycles

Reply:
Ah ok. You should adjust the doc string then.

Reply:
Sorry, you're right, I forgot that was there